### PR TITLE
Direct Support for wait Option for ArgoCD Applications in Kube-Burner

### LINKED ISSUE #691: [RFE]Direct Support for wait Option for ArgoCD Applications in Kube-Burner
Background:

In the context of testing the performance of ArgoCD, I have been using kube-burner to create multiple ArgoCD applications. Each application deploys multiple resources across different namespaces. For accurate performance measurement, it's crucial that kube-burner waits for each ArgoCD application to reach the Synced state before capturing the relevant metrics.

Current Limitation:

While kube-burner supports waiting for specific conditions on child resources, it does not currently support a direct wait option for ArgoCD applications. This limitation arises because the status structure of kube-burner does not match the ArgoCD application status structure. When kube-burner attempts to decode the ArgoCD application status, it results in an empty structure, making it impossible to accurately wait for the Synced state of ArgoCD applications.

Issue #653 Context:

The recent fix/closure of issue #653 addressed the mismatch in the status structure, which now allows kube-burner to watch and wait for child resources created by ArgoCD. However, direct support for waiting on the ArgoCD application status itself is still missing.

Proposed Enhancement:

I propose adding a direct wait option for ArgoCD applications in kube-burner. This enhancement would allow users to specify that kube-burner should wait until an ArgoCD application reaches the Synced state before proceeding. This feature would be particularly useful for performance testing and monitoring, ensuring that all metrics are captured after the ArgoCD applications are fully synced.

Suggested Implementation:

Wait Option for Create Jobs: Enhance the create job wait options to support waiting directly on ArgoCD applications. An example configuration might look like this:
```
objects:
- kind: Application
  objectTemplate: argocd/template/application.yaml
  replicas: 1 
  waitOptions:
    kind: Application
    labelSelector: {app.kubernetes.io/instance: my-argocd-app}
    forCondition: Synced
```

### PR DESCRIPTION
**Pull request by Mangaal**
## Type of change
- [ ] Refactor
-  New feature
- [ ] Bug fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This pull request introduces a new feature to the verifyCondition function, enabling support for jq path expressions. This enhancement allows for flexible and dynamic status checks of any resource by leveraging jq path for querying complex JSON structures

- Added Support for jq Path Expressions: The verifyCondition function now supports jq path expressions for status checks.
- Flexible Status Checks: You can now specify a single jq path string to query nested status fields in CRDs.
- Updated Configuration: The configuration now accepts jq expressions for more dynamic status verification.

## Example Usage:
To use the new feature, you can specify a jq path expression in the customStatusPath field. For example:

```
objects:
  - kind: Application
    objectTemplate: kube-burner-config/template/application.yaml
    replicas: 1
    waitOptions:
      forCondition: "Synced"
      customStatusPath: ".operationState.syncResult.resources[].status"
```

In this example, the system checks if the status field in the syncResult.resources array is Synced.

## Related Tickets & Documents

- Closes #691

### CODE REVIEW

File: pkg/burner/waiters.go [hunk 1]
```diff
@@ -202,7 +203,7 @@ func waitForPod(ns string, maxWaitTimeout time.Duration, obj object, limiter *ra
 	wait.PollUntilContextTimeout(context.TODO(), time.Second, maxWaitTimeout, true, func(ctx context.Context) (done bool, err error) {
 		// We need to paginate these requests to ensure we don't miss any pods
 		listOptions := metav1.ListOptions{
-			Limit: 1000,
+			Limit:         1000,
```
Comment (anandf) [inline pkg/burner/waiters.go:206]: why this extra tab ? check your formatting config.

File: pkg/burner/waiters.go [hunk 2]
```diff
@@ -53,30 +54,30 @@ func (ex *Executor) waitForObjects(ns string, limiter *rate.Limiter) {
...
-					waitForVM(waitNs, ex.MaxWaitTimeout, obj, limiter)
-				case "VirtualMachineInstance":
-					waitForVMI(waitNs, ex.MaxWaitTimeout, obj, limiter)
-				case "VirtualMachineInstanceReplicaSet":
-					waitForVMIRS(waitNs, ex.MaxWaitTimeout, obj, limiter)
-				case "Job":
-					waitForJob(waitNs, ex.MaxWaitTimeout, obj, limiter)
-				case "PersistentVolumeClaim":
-					waitForPVC(waitNs, ex.MaxWaitTimeout, obj, limiter)
+			case "Deployment":
```
Comment (anandf) [inline pkg/burner/waiters.go:57]: There are some formatting changes. Are they really required or is it caused due to some formating `gofmt` misconfiguration.

File: pkg/config/types.go [hunk 3]
```diff
@@ -189,6 +189,8 @@ type WaitOptions struct {
 	LabelSelector map[string]string `yaml:"labelSelector" json:"labelSelector,omitempty"`
 	// ForCondition wait for this condition to become true
 	ForCondition string `yaml:"forCondition" json:"forCondition,omitempty"`
+	// CustomStatusPath defines the path to the specific status field to check (e.g., [ "health", "status"]).
```
Comment (anandf) [inline pkg/config/types.go:192]: This should be a jquery path right ? check the example and correct the documentation.

Comment (Mangaal) [reply pkg/config/types.go]: Yes I have updated the comment

File: pkg/burner/waiters.go (hunk 2 above)
Comment (rsevilla87) [reply pkg/burner/waiters.go:57]: There was an issue with the CI linters, they're going to be fixed at https://github.com/kube-burner/kube-burner/pull/689

Comment (vishnuchalla) [reply pkg/burner/waiters.go:57]: @Mangaal Can you rebase your changes with latest upstream and only make edits where needed? That should fix these formatting issues.

**Comment by vishnuchalla** [comment]
Changes look good to me. @Mangaal would you mind adding a small [functionality test](https://github.com/kube-burner/kube-burner/blob/main/test/test-k8s.bats) for one of your argo CD's usecase so that we always ensure that none of the future changes to this part of code break it?

**Comment by Mangaal** [comment]
Hi @vishnuchalla , I have added the test for the new wait option

File: test/k8s/kube-burner-wait-for-condition.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+global:
```
Comment (vishnuchalla) [inline test/k8s/kube-burner-wait-for-condition.yaml:1]: Adding a new file altogether is not required. Can you just reuse this [configuration](https://github.com/kube-burner/kube-burner/blob/1964a66361bbcf0e24a53af986e76cac225d570a/test/k8s/kube-burner.yml)?

Comment (Mangaal) [reply test/k8s/kube-burner-wait-for-condition.yaml]: I have reused the previous configuration

File: pkg/config/types.go [hunk 5]
```diff
@@ -189,6 +189,8 @@ type WaitOptions struct {
 	LabelSelector map[string]string `yaml:"labelSelector" json:"labelSelector,omitempty"`
 	// ForCondition wait for this condition to become true
 	ForCondition string `yaml:"forCondition" json:"forCondition,omitempty"`
+	// CustomStatusPath defines the jq path to the specific status field to check (e.g., sync.status).
+	CustomStatusPath string `yaml:"customStatusPath" json:"customStatusPath,omitempty"`
```
Comment (rsevilla87) [inline pkg/config/types.go:193]: Can you please add this option to the docs?

**Comment by rsevilla87** [comment]
Overall, I really like this PR :), thanks!, let's document this new option before merging

**Review by rsevilla87 — APPROVED** [review]
lgtm, thanks!

**Pull request merged**